### PR TITLE
Lukšan and Vlček's constrained problems

### PR DESCRIPTION
**Pull request by abelsiqueira**
From [this](https://asep.lib.cas.cz/arl-cav/cs/csg/?repo=crepo1&key=8288792967)
```
V767-98 : Ladislav Luksan, Jan Vlcek:
Sparse and partially separable test problems for unconstrained and equality constrained optimization.
Tech. Rep. V-767, ICS AS CR, December 1998
```
@dpo What do you think of the problem names?

Another problem, relevant for large problems, is that `ADNLSModel` only returns dense matrices. However, the current alternative is to pass all functions and derivatives by hand.

### CODE REVIEW

**Comment by dpo** [comment]
The name is fine. That's a great source of large-scale NLS problems!

Regarding sparse derivatives, I wonder if it would be possible to evaluate derivatives of individual `@NLexpression`s in JuMP. Or somehow extend JuMP to NLS objective terms by (essentially) copying the code used for evaluating and differentiating constraints.

**Comment by abelsiqueira** [comment]
@mlubin, do you know if this is hard to do? Is it better to try to use `ReverseDiffSparse` directly?

**Comment by mlubin** [comment]
Don't use `ReverseDiffSparse` directly. The code has been merged into JuMP, and the standalone package is deprecated. I think it would be reasonably straightforward to create a JuMP extension where:

1. The user provides a collection of `NLexpression`s.
2. The extension creates a new internal JuMP model where those expressions appear as constraints.
3. The extension then calls the jacobian evaluation code and transparently gives the right derivatives to the NLS solver.

This wouldn't require touching any of the AD code.

**Comment by abelsiqueira** [comment]
Miles, thank you for the input.

I'll try to implement the NLS model this was then.

**Comment by dpo** [comment]
Thanks Miles! I think that's the way forward.

**Comment by abelsiqueira** [comment]
So, what I tried making work was
```julia
m = Model()
@variable(m, x[1:2])
@NLexpression(m, F1, 10 * (x[2] - x[1]^2))
@NLexpression(m, F2, x[1] - 1.0)
@NLconstraint(m, x[1]^3 - x[2] <= 1.0)

nls = NLSModel(m, [F1, F2])
```
Internally, I'm thinking we would have two JuMP models, one for the NLS residual function as constraints, and another for the constraints. From what I figured a need something equivalent to
```julia
function NLSModel(m, F)
  Fmodel = Model()
  # Copy variables
  for Fi in F
    @NLconstraint(m, Fi == 0) # or similar
  end
  ...
```
- Is there a decent way to obtain the variables? I managed to create them (I think) by copying a few things (`colCat`, `colLower`, etc.).
- How do I interpret the expression Fi?

**Comment by abelsiqueira** [comment]
Alternatively, I tried using strings for the residual function
```julia
F = ["10 * (x[2] - x[1]^2)"; "x[1] - 1.0"]
...
for Fi in F
  @eval $(parse("@NLconstraint(Fmodel, $Fi == 0)"))
end
```
But this is failing with bounds error. Any suggestions here?

**Comment by mlubin** [comment]
Please don't use `@eval`, it hurts my eyes!

See http://www.juliaopt.org/JuMP.jl/0.18/nlp.html#raw-expression-input for how to add constraints using Julia expressions. Questions on this API are best asked on discourse.

To extract the subexpressions as Julia expressions, it will take small bit of hacking, but here's an example:
```julia
using JuMP
using MathProgBase

m = Model()
@variable(m, x)
ex = @NLexpression(m, x^2)
# Note: "ex" must appear *somewhere* in the objective or constraints for its julia expression to be generated. For example, you can add a meaningless constraint:
@NLconstraint(m, ex <= 0)
evaluator = JuMP.NLPEvaluator(m)
MathProgBase.initialize(evaluator, [:ExprGraph])
@show evaluator.subexpressions_as_julia_expressions[ex.index]
```
This prints
```
evaluator.subexpressions_as_julia_expressions[ex.index] = :(x[1] ^ 2.0)
```

You can then manipulate that and add it to a new model as you like. The usual caveats hold for JuMP internals, they could break between versions.

**Comment by abelsiqueira** [comment]
Needs https://github.com/JuliaSmoothOptimizers/NLPModels.jl/pull/119 and new NLPModels release.

File: src/LVcon501.jl [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+using JuMP
+
+export LVcon501
+
+"""Lukšan and Vlček's problem 5.1 in NLS format:
+Chained Rosenbrock function with trigonometric-exponential constraints
+"""
+function LVcon501(n :: Int = 20)
+
+  n < 3 && error(": number of variables must be ≥ 3")
```
Comment (dpo) [inline src/LVcon501.jl]: Instead of an error, I'd prefer adjusting `n` automatically and throw a warning.

**Comment by abelsiqueira** [comment]
> Is the idea to convert all models in this repo to the JuMP/NLSModel format?

I think so. Though there is little improvement for the small problems. I already started.

**Comment by dpo** [comment]
JuMP should be added to `test/REQUIRE`:
https://travis-ci.org/JuliaSmoothOptimizers/NLSProblems.jl/jobs/380315136#L494

**Comment by abelsiqueira** [comment]
Should be good to go now. Sorry for the delay.

**Pull request merged**